tests/unit/s2n_protocol_preferences_test.c

```diff
@@ -160,6 +162,137 @@ int main(int argc, char **argv)
+            EXPECT_SUCCESS(s2n_stuffer_growable_alloc(&input, 0));
+            EXPECT_SUCCESS(s2n_stuffer_write_uint8(&input, sizeof(protocol1)));
+            EXPECT_SUCCESS(s2n_stuffer_write_bytes(&input, protocol1, sizeof(protocol1)));
+
+            struct s2n_blob result = { 0 };
+            EXPECT_OK(s2n_protocol_preference_read(&input, &result));
+            EXPECT_EQUAL(result.size, sizeof(protocol1));
+            EXPECT_BYTEARRAY_EQUAL(result.data, protocol1, sizeof(protocol1));
+
+            EXPECT_SUCCESS(s2n_stuffer_free(&input));
+        }
+
+        /* Read what we write */
+        {
+            struct s2n_connection *conn = s2n_connection_new(S2N_CLIENT);
+            EXPECT_NOT_NULL(conn);
+
+            EXPECT_SUCCESS(s2n_connection_append_protocol_preference(conn, protocol1, sizeof(protocol1)));
+            EXPECT_SUCCESS(s2n_client_alpn_extension.send(conn, &conn->handshake.io));
+
+            /* Skip list size */
+            EXPECT_SUCCESS(s2n_stuffer_skip_read(&conn->handshake.io, sizeof(uint16_t)));
+
+            struct s2n_blob result = { 0 };
+            EXPECT_OK(s2n_protocol_preference_read(&conn->handshake.io, &result));
+            EXPECT_EQUAL(result.size, sizeof(protocol1));
+            EXPECT_BYTEARRAY_EQUAL(result.data, protocol1, sizeof(protocol1));
+
+            EXPECT_SUCCESS(s2n_connection_free(conn));
+        }
+    }
+
+    /* s2n_protocol_preferences_contain */
+    {
+        uint8_t protocol3[] = "protocol3";
+        struct s2n_blob target = { 0 };
+        EXPECT_SUCCESS(s2n_blob_init(&target, protocol3, sizeof(protocol3)));
+
+        /* Safety checks */
+        {
+            struct s2n_blob blob = { 0 };
+            bool match = false;
+            EXPECT_ERROR_WITH_ERRNO(s2n_protocol_preferences_contain(NULL, &blob, &match), S2N_ERR_NULL);
+            EXPECT_ERROR_WITH_ERRNO(s2n_protocol_preferences_contain(&blob, NULL, &match), S2N_ERR_NULL);
+            EXPECT_ERROR_WITH_ERRNO(s2n_protocol_preferences_contain(&blob, &blob, NULL), S2N_ERR_NULL);
+        }
+
+        /* No supported protocols */
+        {
+            struct s2n_connection *conn = s2n_connection_new(S2N_SERVER);
+            EXPECT_NOT_NULL(conn);
+
+            bool result = true;
+            EXPECT_OK(s2n_protocol_preferences_contain(&conn->application_protocols_overridden, &target, &result));
+            EXPECT_FALSE(result);
+
+            EXPECT_SUCCESS(s2n_connection_free(conn));
+        }
+
+        /* Error: malformed preferences */
+        {
+            struct s2n_connection *conn = s2n_connection_new(S2N_SERVER);
+            EXPECT_NOT_NULL(conn);
+
+            EXPECT_SUCCESS(s2n_connection_append_protocol_preference(conn, protocol1, sizeof(protocol1)));
+            conn->application_protocols_overridden.size--;
+
+            bool result = true;
+            EXPECT_ERROR(s2n_protocol_preferences_contain(&conn->application_protocols_overridden, &target, &result));
```

Comment:
Same as above: Can we be specific and assert the error_no using EXPECT_ERROR_WITH_ERRNO here so that we know the function is error-ing at the right place due to the wrong size being set?

Reply:
https://github.com/aws/s2n-tls/pull/2609#discussion_r582315206